___

<a href='http://www.pieriandata.com'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Copyright by Pierian Data Inc.</em></center>
<center><em>For more information, visit us at <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

# DataFrames

Throughout the course, most of our data exploration will be done with DataFrames. DataFrames are an extremely powerful tool and a natural extension of the Pandas Series. By definition all a DataFrame is:

**A Pandas DataFrame consists of multiple Pandas Series that share index values.**

## Imports

In [92]:
import numpy as np
import pandas as pd

### Creating a DataFrame from Python Objects

In [93]:
# help(pd.DataFrame)

In [94]:
# Make sure the seed is in the same cell as the random call
# https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do
np.random.seed(101)
mydata = np.random.randint(0, 101, (4, 3))

In [95]:
mydata

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

In [96]:
myindex = ['CA', 'NY', 'AZ', 'TX']

In [97]:
my_data = np.random.randint(0, 10, (4, 3))
index = ["MP", "KT", "SC", "JV"]
columns = ["WO", "LO", "LO"]
my_data_frame = pd.DataFrame(data=my_data, index=index, columns=columns)
my_data_frame

WO  LO  LO
MP   8   5   0
KT   5   8   1
SC   3   8   3
JV   3   2   8

In [98]:
mycolumns = ['Jan', 'Feb', 'Mar']

In [99]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

In [100]:
df = pd.DataFrame(data=mydata, index=myindex)
df

0   1   2
CA  95  11  81
NY  70  63  87
AZ  75   9  77
TX  40   4  63

In [101]:
df = pd.DataFrame(data=mydata, index=myindex, columns=mycolumns)
df

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [102]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int64
 1   Feb     4 non-null      int64
 2   Mar     4 non-null      int64
dtypes: int64(3)
memory usage: 128.0+ bytes



# Reading a .csv file for a DataFrame

----

## NOTE: We will go over all kinds of data inputs and outputs (.html, .csv, .xlxs , etc...) later on in the course! For now we just need to read in a simple .csv file.

----

## CSV
Comma Separated Values files are text files that use commas as field delimeters.<br>
Unless you're running the virtual environment included with the course, you may need to install <tt>xlrd</tt> and <tt>openpyxl</tt>.<br>
In your terminal/command prompt run:

    conda install xlrd
    conda install openpyxl

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

### Understanding File Paths

You have two options when reading a file with pandas:

1. If your .py file or .ipynb notebook is located in the **exact** same folder location as the .csv file you want to read, simply pass in the file name as a string, for example:
    
        df = pd.read_csv('some_file.csv')
        
2. Pass in the entire file path if you are located in a different directory. The file path must be 100% correct in order for this to work. For example:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")

#### Print your current directory file path with pwd

In [103]:
pwd

'/home/krzysztofs/notebooks/03-Pandas'

#### List the files in your current directory with ls

In [104]:
ls

 00-Series.ipynb                              hotel_booking_data.csv
 01-DataFrames.ipynb                          movie_scores.csv
 02-Conditional-Filtering.ipynb               mpg.csv
 03-Useful-Methods.ipynb                      my_excel_file.xlsx
 04-Missing-Data.ipynb                        new_file.csv
 05-Groupby-Operations-and-MultiIndex.ipynb   newfile.csv
 06-Combining-DataFrames.ipynb                new_workbook.xlsx
 07-Text-Methods.ipynb                        __pycache__/
 08-Time-Methods.ipynb                        reshaping_pivot.png
 09-Inputs-and-Outputs.ipynb                  RetailSales_BeerWineLiquor.csv
 10-Pivot-Tables.ipynb                        Sales_Funnel_CRM.csv
'11-Pandas-Project-Exercise .ipynb'           sample_table.html
 12-Pandas-Project-Exercise-Solution.ipynb    simple.html
 example.csv                                  tips.csv
 example.xlsx


In [105]:
df = pd.read_csv('tips.csv')

In [106]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).

* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

## Obtaining Basic Information About DataFrame

In [107]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [108]:
df.head(10)
df.describe().transpose()
df["percentage"] = df["tip"] / df["total_bill"] * 100
df["percentage"] = df["percentage"].round(2)
df


total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.94  
1        Douglas Tucker  4478071379779230    Sun4608       16.05  
2        Travis Walters  6011812112971322    Sun4458       16.66  
3      Nathaniel Harris  4676137647685994    Sun5260       13.98  
4          Tonya Carter  4832732618637221    Sun2251       14.68  
..                  ...               ...        ...         ...  
239       Michael Avila  5296068606052842    Sat2657       20.39  
240      Monica Sanders  3506806155565404    Sat1766        7.36  
241          Keith Wong  6011891618747196    Sat3880        8.82  
242        Dennis Dixon     4375220550950      Sat17        9.82  
243     Michelle Hardin  3511451626698139    Thur672       15.97  

[244 rows x 12 columns]

In [109]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [110]:
df.head(3)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   

           Payer Name         CC Number Payment ID  percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.94  
1      Douglas Tucker  4478071379779230    Sun4608       16.05  
2      Travis Walters  6011812112971322    Sun4458       16.66

In [111]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  percentage  
241       Keith Wong  6011891618747196    Sat3880        8.82  
242     Dennis Dixon     4375220550950      Sat17        9.82  
243  Michelle Hardin  3511451626698139    Thur672       15.97

In [112]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
 11  percentage        244 non-null    float64
dtypes: float64(4), int64(2), object(6)
memory usage: 23.0+ KB


In [113]:
len(df)

244

In [114]:
df.describe()

total_bill         tip        size  price_per_person     CC Number  \
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02   
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15   
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15   
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10   
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13   
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15   
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15   
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15   

       percentage  
count  244.000000  
mean    16.079754  
std      6.107020  
min      3.560000  
25%     12.910000  
50%     15.475000  
75%     19.147500  
max     71.030000

In [115]:
df.describe().transpose()

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   
percentage        244.0  1.607975e+01  6.107020e+00  3.560000e+00   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15  
percentage        1.291000e+01  1.547500e+01  1.914750e+01  7.103000e+01

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [116]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.94  
1      Douglas Tucker  4478071379779230    Sun4608       16.05  
2      Travis Walters  6011812112971322    Sun4458       16.66  
3    Nathaniel Harris  4676137647685994    Sun5260       13.98  
4        Tonya Carter  4832732618637221    Sun2251       14.68

#### Grab a Single Column

In [117]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [118]:
type(df['total_bill'])

pandas.core.series.Series

#### Grab Multiple Columns

In [119]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill', 'tip']]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

In [120]:
df['tip_percentage'] = 100 * df['tip'] / df['total_bill']

In [121]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  percentage  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.94        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.05       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.66       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.98       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.68       14.680765

In [122]:
df['price_per_person'] = df['total_bill'] / df['size']

In [123]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.66    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.50    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.31    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.61  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  percentage  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.94        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.05       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.66       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.98       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.68       14.680765

In [124]:
help(np.round)

Help on _ArrayFunctionDispatcher in module numpy:

round(a, decimals=0, out=None)
    Evenly round to the given number of decimals.
    
    Parameters
    ----------
    a : array_like
        Input data.
    decimals : int, optional
        Number of decimal places to round to (default: 0).  If
        decimals is negative, it specifies the number of positions to
        the left of the decimal point.
    out : ndarray, optional
        Alternative output array in which to place the result. It must have
        the same shape as the expected output, but the type of the output
        values will be cast if necessary. See :ref:`ufuncs-output-type` for more
        details.
    
    Returns
    -------
    rounded_array : ndarray
        An array of the same type as `a`, containing the rounded values.
        Unless `out` was specified, a new array is created.  A reference to
        the result is returned.
    
        The real and imaginary parts of complex numbers are rounded
      

#### Adjust Existing Columns

In [125]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'], 2)

In [126]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  percentage  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.94        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.05       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.66       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.98       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.68       14.680765

#### Remove Columns

In [127]:
# df.drop('tip_percentage',axis=1)

In [128]:
df = df.drop("tip_percentage", axis=1)

In [129]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.94  
1      Douglas Tucker  4478071379779230    Sun4608       16.05  
2      Travis Walters  6011812112971322    Sun4458       16.66  
3    Nathaniel Harris  4676137647685994    Sun5260       13.98  
4        Tonya Carter  4832732618637221    Sun2251       14.68

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [130]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.94  
1      Douglas Tucker  4478071379779230    Sun4608       16.05  
2      Travis Walters  6011812112971322    Sun4458       16.66  
3    Nathaniel Harris  4676137647685994    Sun5260       13.98  
4        Tonya Carter  4832732618637221    Sun2251       14.68

In [131]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [132]:
df.set_index('Payment ID')

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  percentage  
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410        5.94  
Sun4608                 3.45      Douglas Tucker  4478071379779230       16.05  
Sun4458                 7.00      Travis Walters  6011812112971322       16.66  
Sun5260                11.84    Nathaniel Harris  4676137647685994       13.98  
Sun2251                 6.15        Tonya Carter  4832732618637221       14.68  
...                      ...                 ...               ...         ...  
Sat2657                 9.68       Michael Avila  5296068606052842       20.39  
Sat1766                13.59      Monica Sanders  3506806155565404        7.36  
Sat3880                11.34          Keith Wong  6011891618747196        8.82  
Sat17                   8.91        Dennis Dixon     4375220550950        9.82  
Thur672                 9.39     Michelle Hardin  3511451626698139       15.97  

[244 rows x 11 columns]

In [133]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.94  
1      Douglas Tucker  4478071379779230    Sun4608       16.05  
2      Travis Walters  6011812112971322    Sun4458       16.66  
3    Nathaniel Harris  4676137647685994    Sun5260       13.98  
4        Tonya Carter  4832732618637221    Sun2251       14.68

In [134]:
df = df.set_index('Payment ID')

In [135]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  percentage  
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410        5.94  
Sun4608                 3.45      Douglas Tucker  4478071379779230       16.05  
Sun4458                 7.00      Travis Walters  6011812112971322       16.66  
Sun5260                11.84    Nathaniel Harris  4676137647685994       13.98  
Sun2251                 6.15        Tonya Carter  4832732618637221       14.68

In [136]:
df = df.reset_index()

In [137]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  percentage  
0              8.49  Christy Cunningham  3560325168603410        5.94  
1              3.45      Douglas Tucker  4478071379779230       16.05  
2              7.00      Travis Walters  6011812112971322       16.66  
3             11.84    Nathaniel Harris  4676137647685994       13.98  
4              6.15        Tonya Carter  4832732618637221       14.68

### ROWS

Let's now explore these same concepts but with Rows.

In [138]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  percentage  
0              8.49  Christy Cunningham  3560325168603410        5.94  
1              3.45      Douglas Tucker  4478071379779230       16.05  
2              7.00      Travis Walters  6011812112971322       16.66  
3             11.84    Nathaniel Harris  4676137647685994       13.98  
4              6.15        Tonya Carter  4832732618637221       14.68

In [139]:
df = df.set_index('Payment ID')

In [140]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  percentage  
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410        5.94  
Sun4608                 3.45      Douglas Tucker  4478071379779230       16.05  
Sun4458                 7.00      Travis Walters  6011812112971322       16.66  
Sun5260                11.84    Nathaniel Harris  4676137647685994       13.98  
Sun2251                 6.15        Tonya Carter  4832732618637221       14.68

#### Grab a Single Row

In [141]:
# Integer Based
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
percentage                        5.94
Name: Sun2959, dtype: object

In [142]:
# Name Based
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
percentage                        5.94
Name: Sun2959, dtype: object

#### Grab Multiple Rows

In [143]:
df.iloc[0:4]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  percentage  
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410        5.94  
Sun4608                 3.45      Douglas Tucker  4478071379779230       16.05  
Sun4458                 7.00      Travis Walters  6011812112971322       16.66  
Sun5260                11.84    Nathaniel Harris  4676137647685994       13.98

In [144]:
df.loc[['Sun2959', 'Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  percentage  
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410        5.94  
Sun5260                11.84    Nathaniel Harris  4676137647685994       13.98

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [145]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  percentage  
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410        5.94  
Sun4608                 3.45      Douglas Tucker  4478071379779230       16.05  
Sun4458                 7.00      Travis Walters  6011812112971322       16.66  
Sun5260                11.84    Nathaniel Harris  4676137647685994       13.98  
Sun2251                 6.15        Tonya Carter  4832732618637221       14.68

In [146]:
df.drop('Sun2959', axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  percentage  
Payment ID                                                                    
Sun4608                 3.45    Douglas Tucker  4478071379779230       16.05  
Sun4458                 7.00    Travis Walters  6011812112971322       16.66  
Sun5260                11.84  Nathaniel Harris  4676137647685994       13.98  
Sun2251                 6.15      Tonya Carter  4832732618637221       14.68  
Sun9679                 6.32        Erik Smith   213140353657882       18.62

In [147]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [148]:
one_row = df.iloc[0]

In [149]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
percentage                        5.94
Name: Sun2959, dtype: object

In [150]:
type(one_row)

pandas.core.series.Series

In [151]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  percentage  
Payment ID                                                                   
Sat2657                 9.68    Michael Avila  5296068606052842       20.39  
Sat1766                13.59   Monica Sanders  3506806155565404        7.36  
Sat3880                11.34       Keith Wong  6011891618747196        8.82  
Sat17                   8.91     Dennis Dixon     4375220550950        9.82  
Thur672                 9.39  Michelle Hardin  3511451626698139       15.97

In [152]:
df.append(one_row).tail()

AttributeError: 'DataFrame' object has no attribute 'append'

--------